### PR TITLE
make the connection ID length configurable

### LINKED ISSUE #1323: use 0 byte connection IDs for the client in IETF QUIC

### LINKED ISSUE #1425: make the connection ID length configurable

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #1425. Fixes #1323. Depends on #1424 and #1436.

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -159,6 +159,13 @@ type Config struct {
 	// This saves 8 bytes in the Public Header in every packet. However, if the IP address of the server changes, the connection cannot be migrated.
 	// Currently only valid for the client.
 	RequestConnectionIDOmission bool
+	// The length of the connection ID in bytes. Only valid for IETF QUIC.
+	// It can be 0, or any value between 4 and 18.
+	// If not set, the interpretation depends on where the Config is used:
+	// If used for dialing to an address, a 0 byte connection ID will be used.
+	// If used for a server, or dialing on a packet conn, a 4 byte connection ID will be used.
+	// When dialing on a packet conn, the ConnectionIDLength value must be the same for every Dial call.
```
Comment (marten-seemann) [inline interface.go]: @lucas-clemente: Do you think we need a check for this?

**Comment by marten-seemann** [comment]
This PR needs to be fixed. Please don't review yet.

**Comment by marten-seemann** [comment]
@lucas-clemente: This PR is ready for review now.

File: interface.go [hunk 2]
```diff
@@ -165,6 +165,13 @@ type Config struct {
 	// This saves 8 bytes in the Public Header in every packet. However, if the IP address of the server changes, the connection cannot be migrated.
 	// Currently only valid for the client.
 	RequestConnectionIDOmission bool
+	// The length of the connection ID in bytes. Only valid for IETF QUIC.
+	// It can be 0, or any value between 4 and 18.
+	// If not set, the interpretation depends on where the Config is used:
+	// If used for dialing to an address, a 0 byte connection ID will be used.
```
Comment (lucas-clemente) [inline interface.go]: -to

**Pull request merged**